File: dawn_go/yfast_trie.go
package yfasttrie

type node struct {
  val   interface{}
  left  *node
  right *node
}

type Trie struct {
  root *node
}

func New() *Trie {
  return &Trie{nil}
}

func (t *Trie) Add(x interface{}, xs []interface{}) {
  t.root = add(t.root, x, xs)
}

func add(n *node, x interface{}, xs []interface{}) *node {
  if n == nil {
    if len(xs) == 0 {
      return &node{x, nil, nil}
    }
    y := xs[0]
    n = &node{y, nil, nil}
    n.left = add(n.left, x, xs[1:])
    return n
  }
  if x == n.val {
    return n
  }
  if x < n.val {
    n.left = add(n.left, x, xs)
  } else {
    n.right = add(n.right, x, xs[1:])
  }
  return n
}

func (t *Trie) Lookup(xs []interface{}) (interface{}, bool) {
  return lookup(t.root, xs)
}

func lookup(n *node, xs []interface{}) (interface{}, bool) {
  if n == nil {
    return nil, false
  }
  if len(xs) == 0 {
    return n.val, true
  }
  y := xs[0]
  if y < n.val {
    return lookup(n.left, xs)
  }
  return lookup(n.right, xs[1:])
}

File: dawn_python/yfast_trie.py
class YFastTrie:
    class Node:
        def __init__(self, val, left=None, right=None):
            self.val = val
            self.left = left
            self.right = right

    def __init__(self):
        self.root = None

    def add(self, x, xs):
        self.root = self._add(self.root, x, xs)

    def _add(self, node, x, xs):
        if node is None:
            if not xs:
                return YFastTrie.Node(x)
            y = xs[0]
            new_node = YFastTrie.Node(y)
            new_node.left = self._add(None, x, xs[1:])
            return new_node
        if x == node.val:
            return node
        if x < node.val:
            node.left = self._add(node.left, x, xs)
        else:
            node.right = self._add(node.right, x, xs[1:])
        return node

    def lookup(self, xs):
        return self._lookup(self.root, xs)

    def _lookup(self, node, xs):
        if node is None:
            return None
        if not xs:
            return node.val
        y = xs[0]
        if y < node.val:
            return self._lookup(node.left, xs)
        else:
            return self._lookup(node.right, xs[1:])